File: applications/rtki0estimation/rtki0estimation.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtki0estimation_ggo.h"
#include "rtkMacro.h"
#include "rtkGgoFunctions.h"

#include <itkExtractImageFilter.h>
#include "rtkI0EstimationProjectionFilter.h"
#include "rtkProjectionsReader.h"

#include <vector>
#include <algorithm>
#include <string>

int main(int argc, char *argv[])
{
  GGO(rtki0estimation, args_info);

  typedef unsigned short InputPixelType;
  const unsigned int Dimension = 3;
  typedef itk::Image< InputPixelType, Dimension > InputImageType;
  typedef itk::Image< unsigned, Dimension >       OutputHistogramType;

  typedef rtk::ProjectionsReader< InputImageType > ReaderType;
  ReaderType::Pointer reader = ReaderType::New();
  reader->SetFileNames( rtk::GetProjectionsFileNamesFromGgo(args_info) );
  TRY_AND_EXIT_ON_ITK_EXCEPTION( reader->UpdateOutputInformation() )

  typedef itk::ExtractImageFilter< InputImageType, InputImageType > ExtractFilterType;
  ExtractFilterType::Pointer extract = ExtractFilterType::New();
  extract->InPlaceOff();
  extract->SetDirectionCollapseToSubmatrix();
  extract->SetInput( reader->GetOutput() );

  ExtractFilterType::InputImageRegionType subsetRegion = reader->GetOutput()->GetLargestPossibleRegion();
  subsetRegion = reader->GetOutput()->GetLargestPossibleRegion();
  extract->SetInput( reader->GetOutput() );
  InputImageType::SizeType extractSize = subsetRegion.GetSize();
  extractSize[2] = 1;
  InputImageType::IndexType start = subsetRegion.GetIndex();

  typedef rtk::I0EstimationProjectionFilter< InputImageType, InputImageType, 2 > I0FilterType;

  std::vector< unsigned short > I0buffer;

  int istep         = 1;
  unsigned int imin = 1;
  unsigned int imax = (unsigned int)subsetRegion.GetSize()[2];
  if ( args_info.range_given )
    {
    if ( ( args_info.range_arg[0] <= args_info.range_arg[2] )
         && ( istep <= ( args_info.range_arg[2] - args_info.range_arg[0] ) ) )
      {
      imin = args_info.range_arg[0];
      istep = args_info.range_arg[1];
      imax = std::min(unsigned(args_info.range_arg[2]), imax);
      }
    }

  I0FilterType::Pointer i0est = I0FilterType::New();

  if ( args_info.lambda_given )
    {
    i0est->SetLambda(args_info.lambda_arg);
    }
  if ( args_info.expected_arg != 65535 )
    {
    i0est->SetExpectedI0(args_info.expected_arg);
    }
  i0est->SaveHistogramsOn();

  for ( unsigned int i = imin; i < imax; i += istep )
    {
    i0est->SetInput( extract->GetOutput() );

    start[2] = i;
    InputImageType::RegionType desiredRegion(start, extractSize);
    extract->SetExtractionRegion(desiredRegion);

    try
      {
      TRY_AND_EXIT_ON_ITK_EXCEPTION( i0est->UpdateLargestPossibleRegion() )
      }
    catch ( itk::ExceptionObject & err )
      {
      std::cerr << "ExceptionObject caught !" << std::endl;
      std::cerr << err << std::endl;
      return EXIT_FAILURE;
      }

    I0buffer.push_back( i0est->GetI0() );
    I0buffer.push_back( i0est->GetI0rls() );
    I0buffer.push_back( i0est->GetI0fwhm() );
    }

  if ( args_info.debug_given )
    {
    std::ofstream paramFile;
    paramFile.open(args_info.debug_arg);
    std::vector< unsigned short >::const_iterator it = I0buffer.begin();
    for (; it != I0buffer.end(); ++it )
      {
      paramFile << *it << ",";
      }
    paramFile.close();
    }

  return EXIT_SUCCESS;
}

File: applications/rtki0estimation/rtki0estimation.py
#!/usr/bin/env python3
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(description="Estimate I0 from projections")

    parser.add_argument(
        "--verbose", "-v", action="store_true", help="Verbose execution"
    )
    parser.add_argument("--debug", "-d", help="Debug CSV output file name")
    parser.add_argument(
        "--range",
        nargs=3,
        type=int,
        help="Range of projection to analyse min step max",
        metavar=("MIN", "STEP", "MAX"),
    )

    parser.add_argument(
        "--lambda",
        "-l",
        dest="lambda_",
        type=float,
        default=0.8,
        help="RLS estimate coefficient",
    )
    parser.add_argument(
        "--expected",
        "-e",
        dest="expected",
        type=int,
        default=65535,
        help="Expected I0 value",
    )

    rtk.add_rtkinputprojections_group(parser)

    return parser


def process(args_info: argparse.Namespace):
    Dimension = 3
    InputImageType = itk.Image[itk.F, Dimension]
    USImageType = itk.Image[itk.US, Dimension]

    reader = rtk.ProjectionsReader[InputImageType].New()
    rtk.SetProjectionsReaderFromArgParse(reader, args_info)
    reader.SetFileNames(rtk.GetProjectionsFileNamesFromArgParse(args_info))
    reader.UpdateOutputInformation()

    subsetRegion = reader.GetOutput().GetLargestPossibleRegion()

    ExtractFilterType = itk.ExtractImageFilter[InputImageType, InputImageType]
    extract = ExtractFilterType.New()
    extract.InPlaceOff()
    extract.SetDirectionCollapseToSubmatrix()
    extract.SetInput(reader.GetOutput())

    extractSize = list(subsetRegion.GetSize())
    extractSize[2] = 1

    istep = 1
    imin = 1
    imax = subsetRegion.GetSize()[2]
    if args_info.range is not None:
        rmin, rstep, rmax = args_info.range
        if (rmin <= rmax) and (istep <= (rmax - rmin)):
            imin = rmin
            istep = rstep
            imax = min(rmax, imax)

    i0est = rtk.I0EstimationProjectionFilter[USImageType, USImageType, 2].New()
    if args_info.lambda_ is not None:
        i0est.SetLambda(args_info.lambda_)
    if args_info.expected is not None and args_info.expected != 65535:
        i0est.SetExpectedI0(args_info.expected)
    i0est.SaveHistogramsOn()

    I0buffer = []

    CastFilterType = itk.CastImageFilter[InputImageType, USImageType]
    cast = CastFilterType.New()
    cast.SetInput(extract.GetOutput())

    start = list(subsetRegion.GetIndex())
    for i in range(imin, imax, istep):
        start[2] = i
        desiredRegion = itk.ImageRegion[Dimension]()
        desiredRegion.SetIndex(start)
        desiredRegion.SetSize(extractSize)

        extract.SetExtractionRegion(desiredRegion)

        i0est.SetInput(cast.GetOutput())
        i0est.UpdateLargestPossibleRegion()

        I0buffer.append(i0est.GetI0())
        I0buffer.append(i0est.GetI0rls())
        I0buffer.append(i0est.GetI0fwhm())

    if args_info.debug:
        with open(args_info.debug, "w") as f:
            f.write(",".join(str(v) for v in I0buffer))

    return 0


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    return process(args_info)


if __name__ == "__main__":
    main()
